### ISSUE TITLE
The same training set, the same various parameters, using SparkXGBoost training out of the model file is different.

### ISSUE THREAD

**Issue by Pythoneryeah** [dmlc/xgboost#10446]
**env：**
spark：3.2.1
xgb:2.0.3
python:3.8

**code：**

`import numpy as np
from pyspark.sql import SparkSession
from pyspark.sql.types import StructType, StructField, FloatType
from pyspark.ml import Pipeline
from pyspark.ml.feature import VectorAssembler
from pyspark.ml.evaluation import MulticlassClassificationEvaluator
from xgboost.spark import SparkXGBClassifier

spark = SparkSession.builder \
    .appName("XGBoost with PySpark Pipeline") \
    .getOrCreate()

np.random.seed(42)
X = np.random.rand(100000, 10).astype(float) 
y = np.random.randint(2, size=100000).astype(float)  

schema = StructType([StructField("label", FloatType(), False)] +
                    [StructField(f"feature_{j}", FloatType(), False) for j in range(10)])

data = spark.createDataFrame(
    [(float(y[i]),) + tuple(float(x) for x in X[i]) for i in range(100000)],
    schema=schema
)

feature_cols = [f"feature_{j}" for j in range(10)]
assembler = VectorAssembler(inputCols=feature_cols, outputCol="features")

xgb_classifier = SparkXGBClassifier(
    max_depth=5,
    eta=0.1,
    objective='binary:logistic',
    num_round=10
)`

**Run the following code multiple times and get a different model.json：**

`pipeline = Pipeline(stages=[assembler, xgb_classifier])

train_data, test_data = data.randomSplit([0.8, 0.2], seed=42)

model = pipeline.fit(train_data)`

**model comparison**
![image](https://github.com/dmlc/xgboost/assets/37642828/ee31b768-67bd-4aa8-8e7a-9a76832b036e)

**Comment by Pythoneryeah** [comment]
Look at this problem, guys. What about the inconsistent results from SparkXgboost? Is there a way to implement distributed training in python

**Comment by trivialfis** [comment]
That's expected, XGBoost is sensitive to the ordering of samples.

**Comment by trivialfis** [comment]
Distributed frameworks usually arrange the samples in a specific way that's not sequential. The behavior can lead to different model outputs.

Feel free to reopen if there are further questions.

**Comment by darshanik** [comment]
@trivialfis @Pythoneryeah 

Facing the same issue. any other workaround to achieve consistency? 

Because if the model is not able to produce the consistent and same results despite the input features and training data are unchanged then won't it be highly unstable and difficult to compare its older version with the new version?

I tried the same on multiple datasets from kaggle aswell.

Initially I thought its the spark that is not giving out the consistent partitions possibly due to shuffling. But I checked and compared the data frame on multiple trials. I have even ordered the data frame on specific columns like "trans_num" or "id". 

<img width="1097" height="996" alt="Image" src="https://github.com/user-attachments/assets/22181659-b9f7-4118-a537-ce225c6dd1ea" />

<img width="650" height="1031" alt="Image" src="https://github.com/user-attachments/assets/77cd79f0-6dfa-491b-9923-49310ebe5c0a" />

<img width="1510" height="1145" alt="Image" src="https://github.com/user-attachments/assets/7fa73cc8-384b-4b74-a143-5e1e4af601d8" />

Would really appreciate if there is any workaround. Or am I missing something?
